libp2p/daemon/daemonapi.nim

```diff
@@ -155,320 +148,330 @@ type
+    disconnectProperties {.fieldNumber: 7.}: Option[SerializableProperties]
+    psProperties {.fieldNumber: 8.}: Option[SerializablePSProperties]
+
+  SerializableError = object
+    msg {.fieldNumber: 1.}: string
+
+  SerializableStreamResponse = object
+    peer {.fieldNumber: 1, dontOmit.}: seq[byte]
+    address {.fieldNumber: 2, dontOmit.}: seq[byte]
+    proto {.fieldNumber: 3, dontOmit.}: string
+
+  SerializableDHTResponse = object
+    kind {.fieldNumber: 1, pint.}: DHTResponseType
+    peer {.fieldNumber: 2.}: Option[SerializableProperties]
+    value {.fieldNumber: 3.}: seq[byte]
+
+  SerializableTopicsResponse = object
+    topics {.fieldNumber: 1.}: seq[string]
+    peerIDs {.fieldNumber: 2.}: seq[seq[byte]]
+
+  SerializablePubSubResponse = object
+    peer {.fieldNumber: 1.}: seq[byte]
+    data {.fieldNumber: 2.}: seq[byte]
+    seqno {.fieldNumber: 3.}: seq[byte]
+    topics {.fieldNumber: 4.}: seq[string]
+    signature {.fieldNumber: 5.}: seq[byte]
+    key {.fieldNumber: 6.}: seq[byte]
+
+  SerializableResponse = object
+    kind {.fieldNumber: 1, pint.}: ResponseKind
+    error {.fieldNumber: 2.}: Option[SerializableError]
+    streamInfo {.fieldNumber: 3.}: Option[SerializableStreamResponse]
+    identify {.fieldNumber: 4.}: Option[SerializableProperties]
+    dht {.fieldNumber: 5.}: Option[SerializableDHTResponse]
+    peers {.fieldNumber: 6.}: seq[SerializableProperties]
+    ps {.fieldNumber: 7.}: Option[SerializableTopicsResponse]
+
 var daemonsCount {.threadvar.}: int
 
-proc requestIdentity(): ProtoBuffer =
+proc requestIdentity(): seq[byte] =
   ## https://github.com/libp2p/go-libp2p-daemon/blob/master/conn.go
   ## Processing function `doIdentify(req *pb.Request)`.
-  result = initProtoBuffer({WithVarintLength})
-  result.write(initProtoField(1, cast[uint](RequestType.IDENTITY)))
-  result.finish()
+  Protobuf.encode(SerializableRequest(
+    reqType: RequestType.IDENTITY
+  ), {VarIntLengthPrefix})
 
-proc requestConnect(peerid: PeerID,
+proc requestConnect(peer: PeerID,
                     addresses: openarray[MultiAddress],
-                    timeout = 0): ProtoBuffer =
+                    timeout = 0): seq[byte] =
   ## https://github.com/libp2p/go-libp2p-daemon/blob/master/conn.go
   ## Processing function `doConnect(req *pb.Request)`.
-  result = initProtoBuffer({WithVarintLength})
-  var msg = initProtoBuffer()
-  msg.write(initProtoField(1, peerid))
-  for item in addresses:
-    msg.write(initProtoField(2, item.data.buffer))
-  if timeout > 0:
-    msg.write(initProtoField(3, hint64(timeout)))
-  result.write(initProtoField(1, cast[uint](RequestType.CONNECT)))
-  result.write(initProtoField(2, msg))
-  result.finish()
-
-proc requestDisconnect(peerid: PeerID): ProtoBuffer =
+  var byteAddresses: seq[seq[byte]] = newSeq[seq[byte]](addresses.len)
+  for a in 0 ..< addresses.len:
+    byteAddresses[a] = addresses[a].data.buffer
+  Protobuf.encode(SerializableRequest(
```

Comment:
nicer here would be something like (not exact syntax, just for the idea):
```
Protobuf.encode(SerializableRequest{...})
  .reqType(..)
  .connectionProperties(proc (x: auto): 
    x.peerOrAddress(...)
)
```

protobuf is "naturally" streamable on writing (except the length prefix) - requiring the construction of the entire object up-front, in-memory just to write a message is quite the hurdle..

Reply:
Right. So there is a writeField API, which I debated using, except it loses most of the benefits of this library and instead mimics the existing API.

The main problem with porting the existing type definitions is the way values are serialized. Addresses are serialized as a seq[byte], but in reality, they're an Object(Object(seq[byte)). This can be... solved, by using an inline pragma of sorts. That is, don't construct the outer object, but treat its children as immediate children. That requires some... field number trickery, but is possible.

That said, then we restart the whole discussion about what pragmas we want to include and how to keep this library concise.

I also would like to note, although this uses Protobuf.encode from nim-serialization, the API does fully expose the streams. It may be possible to work out a cast based solution, where we cast to equivalent objects, yet with their own pragmas, so we can serialize the start of the object, then manually serialize the addresses in the middle, and then finish with the end. That is a whole other beast though.

Reply:
well, the thing is that writing a few fields of protobuf should be a completely trivial thing to do - the object definitions serve as canonical sources of truth in this case for the field number and type, but beyond that, it's a very high mental tax to pay to bring in hundreds of lines of code just to... write a few bytes to an array - the previous code is a lot more digestible from this point of view.

Instead of being something trivial to audit, analyse and debug, it becomes an black box that does magic that is completely opaque and unrelated to the general notion of "put these key-values in this seq".

Reply:
the thing here is that the data we're trying to encode does not exist as a single object to begin with - when you're reading, you can't really do much better than putting the data in an intermediary "transport" object because of out-of-order fields and last-field-overwrites etc - but that's not really the case when writing.

that said, you won't do much better with a streaming api: because of the varint length prefix, you must know the length of the serialized data and either you have to make two passes or buffer everything in a seq - since most protobufs are small, it's customary to go with the latter (even though more advanced frameworks will have support for the former as well)

Reply:
The "black box" solution has an advantage though. The auditing can be considered "modular". One auditing team/effort can verify that the ProtoBuf library implements the rules of the format correctly and another team/effort can verify that LibP2P is using ProtoBuf as specified.

With that said, the `writeField` set of APIs in nim-serialization exists precisely because constructing  a whole object is sometimes undesirable. The P2P protocol layer and Chronicles for example simulate the writing of an entire object by writing its corresponding fields one by one (the fields are accepted as parameters to the RPC procs). It's possible to define some high-level syntax sugar APIs in nim-serialization to make this more accessible to the user code.

Regarding streamed writing, this is handled efficiently in the library due to the support for "delayed writes" in FastStreams. This solves the problem of varint length prefixes without buffering to a sequence or introducing a "measuring pass".

Reply:
As zah said, delayed writes solve that problem.

For some reason, I completely forgot fields can be unordered. If you mark those arrays as don't serialize, create a stream and call newProtobufWriter with it, manually doing the delayed write + prefix, you can use the writeField internal API to finish it off. Then you don't need any casts.

That said, while you can write like that, you can't read like that. It'd require iterating over the entire stream to find the field, or at least parsing it into a buffer table of `int, seq[byte]`. As this library doesn't iterate over the entire stream, nor provide a buffer table, you'll still need a wrapper object to read into.

Reply:
efficiently, how?

Reply:
The buffering mechanism in FastStreams divides the stream in "pages". A page is allocated and written to only once, but you can create multiple "fat pointers" to the same page. Each such pointer represents a sub-span of page.

When you do a delayed write, you"split" the page by creating a new fat pointer pointing to the same buffer, but now it has an adjustable start position that will be determined once the delayed write is completed.

Since the fat pointers are stored in a `Deque`, this usually doesn't require any new allocations. In the end, when all delayed writes are completed, the outstanding page spans are flushed to the output device, potentially using some [Vectored I/O](https://en.wikipedia.org/wiki/Vectored_I/O) API. Please note in this setup, every byte is written to exactly once and you never need to copy any buffer.

The same mechanism can be used for implementing a zero-copy multiplexer. The incoming bytes exist in sharable pages. Something like Mplex only needs to determine what are the boundaries between the sub-streams and it pushes the same buffers with adjusted readable spans to the next layer.

In async pipes and channels, the pages can flow from one stream to another or they can be shared between a reader and a writer that only need to adjust two offsets - `writtenTo` and `consumedTo`.

Reply:
ok.. so.. I don't quite understand the difference between between that and a `seq` and a `length_counter` in a local variable (minus the paging) though? 

ie if the final destination is disk or network network, ultimately, you will need to buffer the whole chunk in some temporary location (memory, fancy swap-backed pages with layers of fat pointers, whatever), then write the length, then write the data to the actual sink (with `writev` if you want to be fancy)?

Reply:
There will be difference in the async channel scenario (because you'll then have to copy the sequence to a buffer that will be sent to a different thread).

Another difference will be in the case where your length-prefixed content is smallish. You wouldn't want to send it out to the network immediately, so you'll first write it to the temporary sequence, then to a higher-level buffer in the stream and finally it will go out to the I/O device.

OTOH, larger content written to a temporary sequence will suffer from re-allocations and the buffer copying associated with that.

Maybe it's not clear how writing the length works in the end. In the stream buffers, you've left a "hole" when the delayed write was started. You have specified the maximum size for the length prefix and once you know its final size, you close the hole by writing the bytes in the right place in the stream buffer and then you adjust the page span boundaries in such a way that some bytes remain effectively unused.